src/dist/cache.rs

```diff
@@ -49,42 +62,88 @@ impl ClientToolchainCache {
         let tc_cache_dir = cache_dir.join("tc");
         let cache = Mutex::new(TcCache::new(&tc_cache_dir, cache_size).unwrap());
 
+        let mut custom_toolchain_paths = HashMap::new();
+        for ct in config_custom_toolchains.into_iter() {
+            if custom_toolchain_paths.contains_key(&ct.compiler_executable) {
+                panic!("Multiple toolchains for {:?}", ct.compiler_executable)
+            }
+            let config::CustomToolchain { compiler_executable, archive, archive_compiler_executable } = ct;
+
+            debug!("Registering custom toolchain for {:?}", compiler_executable);
+            let custom_tc = CustomToolchain {
+                archive: archive.clone(),
+                compiler_executable: archive_compiler_executable.clone(),
+            };
+            assert!(custom_toolchain_paths.insert(compiler_executable.clone(), (custom_tc, None)).is_none());
+        }
+        let custom_toolchain_paths = Mutex::new(custom_toolchain_paths);
+
         Self {
             cache_dir,
             cache,
+            custom_toolchains: Mutex::new(HashMap::new()),
+            custom_toolchain_paths,
             // TODO: shouldn't clear on restart, but also should have some
             // form of pruning
             weak_map: Mutex::new(weak_map),
         }
     }
 
     // Get the bytes of a toolchain tar
-    pub fn get_toolchain_cache(&self, key: &str) -> Option<Vec<u8>> {
-        let mut toolchain_reader = match self.cache.lock().unwrap().get(key) {
-            Ok(rdr) => rdr,
-            Err(LruError::FileNotInCache) => return None,
-            Err(e) => panic!("{}", e),
+    // TODO: by this point the toolchain should be known to exist
+    pub fn get_toolchain(&self, tc: &Toolchain) -> Option<Vec<u8>> {
+        // TODO: be more relaxed about path casing and slashes on Windows
```

Comment:
File a followup for this? Maybe just running the paths through `fs::canonicalize` would be good enough.